File: src/server/pacing.test.ts
import { describe, it, expect } from 'vitest';
import { derivePacing, KNOWN_BUCKET_WINDOWS } from './pacing.js';
import type { UsageBucket } from '../shared/types.js';

const hr = 3600 * 1000;

describe('derivePacing', () => {
  it('on_track when utilization roughly matches elapsed fraction', () => {
    const now = 5 * hr;
    const b: UsageBucket = { id: 'five_hour', utilization: 0.40, reset_at: 8 * hr, window_started_at: 3 * hr };
    // elapsed = 2h / 5h = 0.4 → delta 0 → on_track
    const p = derivePacing(b, now);
    expect(p.zone).toBe('on_track');
    expect(Math.abs(p.delta)).toBeLessThan(0.01);
  });

  it('hot when utilization >=10pp above expected', () => {
    const now = 5 * hr;
    const b: UsageBucket = { id: 'five_hour', utilization: 0.60, reset_at: 8 * hr, window_started_at: 3 * hr };
    // elapsed 0.4; utilization 0.6 → +20pp
    const p = derivePacing(b, now);
    expect(p.zone).toBe('hot');
    expect(p.delta).toBeCloseTo(0.20, 2);
  });

  it('chill when utilization >=5pp below expected', () => {
    const now = 5 * hr;
    const b: UsageBucket = { id: 'five_hour', utilization: 0.30, reset_at: 7 * hr, window_started_at: 2 * hr };
    // elapsed 3h / 5h = 0.6; utilization 0.3 → -30pp
    expect(derivePacing(b, now).zone).toBe('chill');
  });

  it('none when inside first 5% of window', () => {
    const now = 5 * hr;
    const b: UsageBucket = { id: 'five_hour', utilization: 0.05, reset_at: 10 * hr, window_started_at: (5 * hr) - (0.1 * hr) };
    expect(derivePacing(b, now).zone).toBe('none');
  });

  it('derives window_started_at for known bucket when null', () => {
    const now = 5 * hr;
    const b: UsageBucket = { id: 'five_hour', utilization: 0.40, reset_at: 8 * hr, window_started_at: null };
    // Derived: 8h - 5h = 3h; elapsed 2h / 5h = 0.4 → on_track
    expect(derivePacing(b, now).zone).toBe('on_track');
  });

  it('none for unknown bucket with no window_started_at', () => {
    const b: UsageBucket = { id: 'mystery', utilization: 0.9, reset_at: 10 * hr, window_started_at: null };
    expect(derivePacing(b, 5 * hr).zone).toBe('none');
  });

  it('KNOWN_BUCKET_WINDOWS has the three documented buckets', () => {
    expect(KNOWN_BUCKET_WINDOWS.five_hour).toBe(5 * hr);
    expect(KNOWN_BUCKET_WINDOWS.weekly).toBe(7 * 24 * hr);
    expect(KNOWN_BUCKET_WINDOWS.sonnet_weekly).toBe(7 * 24 * hr);
  });
});

File: server-elixir/test/fbi/usage/pacing_test.exs
defmodule FBI.Usage.PacingTest do
  use ExUnit.Case, async: true
  alias FBI.Usage.Pacing

  @known_bucket %{
    id: "five_hour",
    utilization: 0.5,
    reset_at: 5 * 3600 * 1000 + 1_000_000,
    window_started_at: 1_000_000
  }

  describe "derive_pacing/2" do
    test "returns :none zone when inside the warm-up window (<5% progress)" do
      bucket = %{@known_bucket | window_started_at: 999_999_500}
      now = 1_000_000_000
      # elapsed = 500ms, duration = 5h -> progress ~ tiny -> none
      assert %{delta: +0.0, zone: :none} = Pacing.derive_pacing(bucket, now)
    end

    test "returns :on_track when utilization matches expected progress" do
      # halfway through window, 50% utilized -> delta = 0
      now = @known_bucket.window_started_at + div(5 * 3600 * 1000, 2)
      bucket = %{@known_bucket | utilization: 0.5}
      %{delta: delta, zone: zone} = Pacing.derive_pacing(bucket, now)
      assert_in_delta delta, 0.0, 0.01
      assert zone == :on_track
    end

    test "returns :chill when utilization trails progress by more than 5%" do
      # halfway through, only 10% utilized -> delta ≈ -0.4 -> chill
      now = @known_bucket.window_started_at + div(5 * 3600 * 1000, 2)
      bucket = %{@known_bucket | utilization: 0.1}
      assert %{zone: :chill} = Pacing.derive_pacing(bucket, now)
    end

    test "returns :hot when utilization exceeds progress by >= 10%" do
      # 10% through, 50% utilized -> delta = +0.4 -> hot
      now = @known_bucket.window_started_at + div(5 * 3600 * 1000, 10)
      bucket = %{@known_bucket | utilization: 0.5}
      assert %{zone: :hot} = Pacing.derive_pacing(bucket, now)
    end

    test "returns :none when reset_at is nil" do
      bucket = %{@known_bucket | reset_at: nil}
      assert %{zone: :none, delta: +0.0} = Pacing.derive_pacing(bucket, 0)
    end

    test "derives window_start from reset_at for known buckets when window_started_at missing" do
      bucket = %{@known_bucket | window_started_at: nil}
      # derived start = reset_at - 5h -> halfway through now = reset_at - 2.5h
      now = bucket.reset_at - div(5 * 3600 * 1000, 2)
      %{delta: delta} = Pacing.derive_pacing(bucket, now)
      assert_in_delta delta, 0.0, 0.01
    end
  end
end
